integration_test/common/common.go

```diff
@@ -29,55 +29,71 @@ type ExpectedMetric struct {
 	// The metric type, for example workload.googleapis.com/apache.current_connections.
 	Type string `yaml:"type"`
 	// The value type, for example INT64.
-	ValueType string `yaml:"value_type"`
+	ValueType string `yaml:"value_type" validate:"oneof=BOOL INT64 DOUBLE STRING DISTRIBUTION"`
 	// The kind, for example GAUGE.
-	Kind string `yaml:"kind"`
+	Kind string `yaml:"kind" validate:"oneof=GAUGE DELTA CUMULATIVE"`
 	// The monitored resource, for example gce_instance.
-	// Currently we only test with gce_instance, so we expect
-	// all expectedMetricsEntries to have gce_instance.
-	MonitoredResource string `yaml:"monitored_resource"`
+	// Currently we only test with gce_instance.
+	MonitoredResource string `yaml:"monitored_resource" validate:"oneof=gce_instance"`
 	// Mapping of expected label keys to value patterns.
 	// Patterns are RE2 regular expressions.
 	Labels map[string]string `yaml:"labels"`
 	// If Optional is true, the test for this metric will be skipped.
-	Optional bool `yaml:"optional,omitempty"`
+	Optional bool `yaml:"optional,omitempty" validate:"excluded_with=Representative"`
 	// Exactly one metric in each expected_metrics.yaml must
 	// have Representative set to true. This metric can be used
 	// to test that the integration is enabled.
-	Representative bool `yaml:"representative,omitempty"`
+	Representative bool `yaml:"representative,omitempty" validate:"excluded_with=Optional"`
 }
 
+type LogFields struct {
+	Name        string `yaml:"name" validate:"required"`
+	ValueRegex  string `yaml:"value_regex,omitempty"`
+	Type        string `yaml:"type" validate:"required"`
+	Description string `yaml:"description" validate:"required"`
+}
+
+type ExpectedLog struct {
+	LogName string      `yaml:"log_name" validate:"required"`
+	Fields  []LogFields `yaml:"fields" validate:"required"`
+}
+
+type MinimumSupportedAgentVersion struct {
+	Logging string `yaml:"logging,omitempty"`
+	Metrics string `yaml:"metrics,omitempty"`
+}
+
+type IntegrationMetadata struct {
+	PublicUrl                    string                       `yaml:"public_url,omitempty"`
```

Comment:
`omitempty` here is for `yaml` so that it doesn't need to serialize empty fields. I added `omitempty` to all of the optional fields to make the new generated yaml more similar to the existing ones but we don't need it anymore if I'm removing them from the diff. I can revert these.

Reply:
My bad. For all times, i was thinking of something different..

I think having omitempty in 'yaml' specifier is fine, as it won't set the field if it's missing from the file.  It doesn't after the e2e test much, as set or not set, golang always has it as empty string in either case. The current test does not use the field for querying if the value shows as empty string: https://github.com/GoogleCloudPlatform/ops-agent/blob/19e1eb67f06bf4497678df9a1f111cd1ca32f279/integration_test/third_party_apps_test.go#L217
But in general i think it's good to have omitempty：）
![image](https://user-images.githubusercontent.com/8354694/163438721-cad371f7-a4d2-430f-8fd4-920f1f61cd3a.png)

Reply:
Updated